lab_4_fill_words_by_ngrams/main.py

```diff
@@ -98,6 +132,38 @@ def run(self, seq_len: int, prompt: str) -> str:  # type: ignore
                 or if sequence has inappropriate length,
                 or if methods used return None.
         """
+        if not (isinstance(prompt, str) and prompt and
+                isinstance(seq_len, int) and seq_len > 0):
+            raise ValueError("TopPGenerator.run: Inappropriate input.")
+
+        encoded = self._word_processor.encode(prompt)
+        if not encoded:
+            raise ValueError("TopPGenerator.run: Couldn't encode prompt.")
+
+        while seq_len >= 1:
+            tokens = self._model.generate_next_token(encoded)
+            if tokens is None:
+                raise ValueError("TopPGenerator.run: "
+                                 "Model returned None instead of generated tokens.")
+            if not tokens:
+                break
+            sorted_tokens = sorted(list(tokens.items()),
+                                   key=lambda pair: (float(pair[1]), pair[0]), reverse=True)
+            sum_probability = 0
+            for index, (_, prob) in enumerate(sorted_tokens):
+                sum_probability += prob
+                if sum_probability >= self._p_value:
+                    random_token = choice(sorted_tokens[:index + 1])
+                    encoded += (random_token[0],)
+                    break
+            else:
+                break
```

Comment:
what does it mean?

Reply:
If the sum of tokens probability is lower that p_value, we basically have no tokens to add, so we stop generation as Step 3.2 requires. Otherwise, we'll be doing the same thing in while loop, untill seq_len is above 0, never adding tokens